core/state/database.go

```diff
@@ -255,23 +256,106 @@ func (tds *TrieDbState) Copy() *TrieDbState {
 
 	if debug.IsIntermediateTrieHash() {
 		cpy.tp.SetUnloadNodeFunc(cpy.putIntermediateHash)
-		cpy.t.SetOnDeleteSubtreeFunc(tds.markSubtreeEmptyInIntermediateHash)
 		cpy.tp.SetCreateNodeFunc(cpy.delIntermediateHash)
 	}
 
 	return &cpy
 }
 
-func (tds *TrieDbState) markSubtreeEmptyInIntermediateHash(prefix []byte) {
-	if len(prefix) != common.HashLength {
+func ClearTombstonesForReCreatedAccount(db ethdb.MinDatabase, addrHash common.Hash) error {
+	addrHashBytes := addrHash[:]
+	if ok, err := HasTombstone(db, addrHashBytes); err != nil {
+		return err
+	} else if !ok {
+		return nil
+	}
+
+	buf := pool.GetBuffer(common.HashLength * 2)
+	defer pool.PutBuffer(buf)
+	copy(buf.B, addrHashBytes)
+
+	i := common.HashLength
+	for j := 0; j < 256; j++ {
+		buf.B[i] = uint8(j)
+		if err := db.Put(dbutils.IntermediateTrieHashBucket, buf.B[:i+1], []byte{}); err != nil {
+			return err
+		}
+	}
+
+	if err := db.Delete(dbutils.IntermediateTrieHashBucket, addrHashBytes); err != nil {
+		return err
+	}
+	return nil
+}
+
+// PutTombstoneForDeletedAccount - placing tombstone only if given account has storage in database
+func PutTombstoneForDeletedAccount(db ethdb.MinDatabase, addrHash []byte) {
+	if len(addrHash) != common.HashLength {
 		return
 	}
 
-	if err := tds.db.Put(dbutils.IntermediateTrieHashBucket, prefix, []byte{}); err != nil {
+	v, _ := db.Get(dbutils.AccountsBucket, addrHash)
+	if v == nil {
+		return
+	}
+	acc := accounts.NewAccount()
+	if err := acc.DecodeForStorage(v); err != nil {
+		log.Warn("could not decode account", "err", err)
+		return
+	}
+
+	if acc.Root == trie.EmptyRoot {
+		return
+	}
+
+	if err := db.Put(dbutils.IntermediateTrieHashBucket, addrHash, []byte{}); err != nil {
 		log.Warn("could not put intermediate trie hash", "err", err)
 	}
 }
 
+func ClearTombstonesForNewStorage(someStorageExistsInThisSubtree func(prefix []byte) bool, db ethdb.MinDatabase, compositeKey []byte) error {
+	if someStorageExistsInThisSubtree(compositeKey) { // this func is only for newly created storage
+		return nil
+	}
+
+	buf := pool.GetBuffer(128)
+	defer pool.PutBuffer(buf)
+
+	buf.B = buf.B[:cap(buf.B)]
+	foundTombStone := false
+	for i := common.HashLength + 1; i < len(compositeKey)-1; i++ { // +1 because first step happened during account re-creation
+		if someStorageExistsInThisSubtree(compositeKey[:i]) {
+			continue
+		}
+
+		if !foundTombStone {
+			if ok, err := HasTombstone(db, compositeKey[:i]); err != nil {
+				return err
+			} else if !ok {
+				continue
+			}
+			foundTombStone = true
+		}
+
+		copy(buf.B, compositeKey[:i+1])
+		for j := 0; j < 256; j++ {
+			if compositeKey[i] == uint8(j) {
```

Comment:
Most of the iterations can be eliminated if we use iterator over existing storage and tombstones